[openmp] Segfaults/assertion errors on certain omp statements after calling `omp_pause_resource_all(omp_pause_hard)`

haiyanghee opened the pull request:
My attempt to fix [#154201](https://github.com/llvm/llvm-project/issues/154201).

I've run the omp tests using the `check-openmp` target, and the new tests pass without segfaults/assertion errors/deadlocks.

However, I'm not certain that my changes are complete (i.e. I didn't miss any edge cases), as I occasionally get 1 or 2 segfaults when I run my changes with my company's regression test that heavily uses omp (I'm uncertain if its my omp changes or company's code base that caused the segfaults). However, I couldn't reproduce any segfaults with the tests I've added.

Also this is my first time contributing to open source, so any feedback is appreciated! :)

Commit:
Fixed mutex segfault after hard reset

Commit:
Prevented double atfork() handler initialization

Commit:
Fixed a deadlock in `#pragma omp parallel num_threads` after hard reset

haiyanghee committed:
Added my test cases to test/api/omp_pause_resource.c

haiyanghee committed:
Do not assert failure if `__kmp_unregister_library()` can't find the shm file because we could have unregistered the library right before fork via hard reset, and child can simply exit immediately.

haiyanghee commented:
Hi @jprotze , would you please be able to review this PR? If not can you please tell me who should review this?

jprotze committed:
Merge branch 'main' into ompHardResetBugFixes-mainBranch

jprotze commented on openmp/runtime/src/kmp_lock.cpp:
```diff
@@ -3431,6 +3431,9 @@ void __kmp_cleanup_indirect_user_locks() {
       __kmp_free(ll->lock);
       ll->lock = NULL;
+      // reset the reverse critical section pointer to 0
+      if (ll->rev_ptr_critSec && LIKELY(!__kmp_in_atexit))
```
```suggestion
      if (ll->rev_ptr_critSec && !UNLIKELY(__kmp_in_atexit))
```
Avoids adding new `LIKELY`

jprotze commented on openmp/runtime/src/z_Linux_util.cpp:
```diff
@@ -1404,13 +1404,15 @@ static void __kmp_atfork_child(void) {
 void __kmp_register_atfork(void) {
-  if (__kmp_need_register_atfork) {
+  // NOTE: we will not double register our fork handlers! It will cause deadlock
+  if (!__kmp_already_registered_atfork && __kmp_need_register_atfork) {
```
Is this change related to the pause_resource_all issue?

haiyanghee replied:
Technically yes, because if you see the test function `test_KMP_INIT_AT_FORK_with_fork_after_omp_hard_pause_resource_all()` I added in the unit tests, the original behaviour will actually register the atfork handlers twice (if the environment variable `KMP_INIT_AT_FORK` is explicitly set to 1), which causes a deadlock (since the atfork handlers are run twice, and inside the handler it will do locking).

jprotze commented on openmp/runtime/src/kmp_runtime.cpp:
```diff
@@ -8343,6 +8345,10 @@ void __kmp_cleanup(void) {
 
   __kmpc_destroy_allocator(KMP_GTID_SHUTDOWN, __kmp_def_allocator);
   __kmp_def_allocator = omp_default_mem_alloc;
+#ifdef KMP_TDATA_GTID
+  /*reset __kmp_gtid to initial value*/
+  __kmp_gtid = KMP_GTID_DNE;
+#endif
```
Thread 0 should not drop it's gtid

jprotze replied:
When you look at code generated for an OpenMP program (IR or disassembled objdump), you will see that each function that contains `__kmpc*` calls initially gets the gtid using `__kmpc_global_thread_num`, which triggers serial initialization. This gtid is then passed to many `__kmpc` calls. You already ran into the issue that `__kmpc_push_num_threads` assumes that serial initialization has happened. I think, that the runtime must be in serial initialized state after omp_pause_resource[_all] to not break the assumptions of compiler+runtime implementation.

haiyanghee replied:
Right I did realize clang inserts `__kmpc_global_thread_num()` at the beginning of functions. 

But how does not resetting `__kmp_gtid` to `KMP_GTID_DNE` for thread 0 imply that the runtime in serial initialized state after omp_pause_resource calls? It sounds like we want to call `__kmpc_global_thread_num()` (or something equivalent) right after omp_pause_resource to force serial initialized state?

jprotze commented on openmp/runtime/src/kmp_csupport.cpp:
```diff
@@ -233,6 +233,11 @@ void __kmpc_push_num_threads(ident_t *loc, kmp_int32 global_tid,
+  // we'll do middle initialize first, as otherwise the assert on global_tid can
+  // fail when omp is not initialized and this function is called
+  if (!TCR_4(__kmp_init_middle)) {
+    __kmp_middle_initialize();
```
I think, kmp_serial_intialize should be sufficient. But I think the primary thread should not drop it's id, see below.

jprotze reviewed (commented):
A first round of comments

haiyanghee commented on openmp/runtime/src/kmp_lock.cpp:
```diff
@@ -3431,6 +3431,9 @@ void __kmp_cleanup_indirect_user_locks() {
       __kmp_free(ll->lock);
       ll->lock = NULL;
+      // reset the reverse critical section pointer to 0
+      if (ll->rev_ptr_critSec && LIKELY(!__kmp_in_atexit))
```
In `__kmp_allocate_indirect_lock()` the new lock tables are allocated with multiples of  `sizeof(kmp_indirect_lock_t)`, and I believe `__kmp_allocate()` will memset the allocated memory to 0. So I believe `rev_ptr_critSec` is technically initialized to 0 in this case?

I can also explicitly initialize it to nullptr if that makes the code easier to reason about.

haiyanghee committed:
removed the LIKELY macro and changed `LIKELY(!__kmp_in_atexit))` to `!UNLIKELY(__kmp_in_atexit))` as suggested since they are equivalent

haiyanghee committed:
call `__kmp_serial_initialize()` instead of `__kmp_middle_initialize()`
in `__kmpc_push_num_threads()`, as suggested in feedback

Commit:
Revert "Fixed a deadlock in `#pragma omp parallel num_threads` after hard reset"

This reverts commit 773ab2f3f49f949553052ea1b84b2c38530489ac.

Commit:
Reuse the reset and re-initialize code in `__kmp_atfork_child()` to `__kmp_hard_pause()`

This should ensure that the program is in serially initialized state after
doing a hard pause, hence not breaking any compiler/runtime
invariants/assumptions.

Also fixed formatting issues

haiyanghee commented:
Hi @jprotze , I attempted to address the issue where after `omp_pause_resource_all()` is called, the runtime must be in serial initialized state in order to not break existing compiler/runtime assumptions in my new commits.

I can simply call `__kmp_serial_initialize()` right after `__kmp_hard_pause()`, but the only issue I worry about is that if we fork right after `__kmp_hard_pause()` and we have `KMP_INIT_AT_FORK=0` set (so we don't run any atfork handlers), the runtime state in the child is unusable. As far as I understand, the omp implementation should make it possible to fork after we hard reset the omp runtime.

So I want the serial initialization to have the following properties:
- (Property 1) does not spawn new pthreads, or any pthread data structure that cannot be used after forking
- (Property 2) does not open and shared memory or shared any undesired file descriptors

And it looks like we should be already doing this inside the `__kmp_atfork_child()` atfork handler. So I basically just did the same thing right after  `__kmp_hard_pause()`.

I tried to understand what `__kmp_do_serial_initialize()` does, but I didn't completely understand why every part of it is needed, and I also don't know all of the compiler and runtime assumptions. So I hope that by doing what the child fork handler is doing I don't break any existing invariants.

Here is the reason why I think the serial initialization satisfies the 2 properties I've stated above:

<details>

I went in gdb and break pointed all of the pthread functions to see what will be called during `__kmp_do_serial_initialize()` by doing `rbreak ^pthread_.*`. 

I see only these pthread functions are used in `__kmp_do_serial_initialize()`:

```
- pthread_once()
    - called in ompt_pre_init()
    - I think this is ok as its just executing things once in ompt (I'm not
familiar with ompt so I hope we don't need to do any clean up up on fork, as
the child fork handler didn't do any)
- pthread_key_create()
    - called in __kmp_runtime_initialize() when we create the thread private keys
    - this is ok, since pthread internally should just malloc some arrays for
thread storage in the future
- pthread_setspecific() and pthread_getspecific()
    - called in `__kmp_register_root(TRUE)` when we call `__kmp_gtid_set_specific()`,
it will write the gtid to the thread private variable
        - NOTE that we didn't call any pthread_create yet
    - this is ok, as the gtid written will always be 0 (line 3863 where we set `gtid=0` in `__kmp_register_root()`). This is because:
        - the input argument `initial_thread` to `__kmp_register_root()` is TRUE
        - and `__kmp_threads[0]` should be 0
        - and we shouldn't have set `__kmp_init_hidden_helper_threads` at that time
            - as it looks like `__kmp_init_hidden_helper_threads` is set to true in `__kmp_task_alloc()`
- pthread_mutex_lock() and pthread_mutex_unlock()
    - called in __kmp_itt_initialize()
    - this is ok, mutex should be reusable after forking
```

So I believe property 1 holds.

Next, I break pointed the `open()` system call to catch any file descriptor openings. The only place where I see we open file descriptors is in `__kmp_register_library_startup()` where we might open a shared memory file. However, since we set `__kmp_need_register_serial()` to false before we call `__kmp_do_serial_initialize()`, it will not call `__kmp_register_library_startup()`.

So I believe property 2 also holds.

**NOTE:** this only verifies that the properties hold for a certain compilation, as other compiler flags might introduce different behaviour.

</details>

Please correct me if my understanding is wrong, as I'd love to know and do this properly.

haiyanghee commented:
Hi @jprotze , can you please check if my attempt at making sure the runtime must be in serial initialized state after calling `omp_pause_resource_all()`? If it doesn't work, can you please guide me to an approach that would work? (maybe the fork handler is the way to go here, and what I'm trying to do is impossible?)
